### ISSUE TITLE
"[Optional]" should not be used with "ref" or "out"

### ISSUE THREAD

**Issue by tvsonar** [SonarSource-VisualStudio/sonarlint-roslyn-analyzer#671]
R# _OptionalParameterRefOut_

You can use optional parameters in C# with the following two methods:
```C#
class MyClass
{
    public void DoStuff([Optional]int i, int j = default(int))  // both i & j optional; specified in 2 different ways
    {
        Console.WriteLine(i);
    }

    public static void Main()
    {
        new MyClass().DoStuff(); // prints 0
        new MyClass().DoStuff(42); // prints 42
    }
}
```
The second option `int j = default(int)` can't be used with `out` or `ref` keywords, compiler error CS1741 is triggered. On the other hand the `OptionalAttribute` can be specified on `ref` and `out` parameters, but shouldn't, because the callers must provide an appropriate argument to the call. If not, CS7036 is triggered.

```C#
class MyClass
{
    public void DoStuff([Optional]ref int i) // Noncompliant, no error from the compiler though
    {
        Console.WriteLine(i);
    }

    public static void Main()
    {
        var i = 24;
        new MyClass().DoStuff(); // This doesn't compile, CS7036 shows
        new MyClass().DoStuff(ref i); // prints 24
    }
}
```

**Comment by ganncamp** [comment]
Maybe combined rule:

"[Optional]" should be used appropriately
* should be last in signature
* not "ref"

**Comment by ganncamp** [comment]
https://jira.sonarsource.com/browse/RSPEC-3447
